File: 121_best_time_buy_sell_stock/best_time_buy_sell_stock.go
package main

import "fmt"

type Solution struct {
}

func (*Solution) maxProfit(prices []int) int {
	min_prices := int(^uint(0) >> 1)
	max_profit := 0
	for _, price := range prices {
		if price < min_prices {
			min_prices = price
		} else if price-min_prices > max_profit {
			max_profit = price - min_prices
		}
	}

	return max_profit
}

func main() {
	solution := new(Solution)
	prices := []int{7, 1, 5, 3, 6, 4}
	result := solution.maxProfit(prices)
	fmt.Println(result)
}

File: 121_best_time_buy_sell_stock/best_time_buy_sell_stock.py
from typing import List

class Solution:
    def max_profit(self, prices: List[int]) -> int:
        min_prices = float('inf')
        max_profit = 0

        for price in prices:
            if price < min_prices:
                min_prices = price
            elif price - min_prices > max_profit:
                max_profit = price - min_prices
            
        return max_profit

if __name__ == '__main__':
    problem = Solution()
    prices = [7,1,5,3,6,4]
    result = problem.calculate(prices)